pretyping/evarconv.ml

```diff
@@ -1110,117 +1211,289 @@ let set_solve_evars f = solve_evars := f
+                                      prc evenv rhs');
+           (** solve_evars is not commuting with nf_evar, because restricting
+               an evar might provide a more specific type. *)
+           let evd, _ = !solve_evars evenv evd rhs' in
+           evdref := evd;
+           if !debug_ho_unification then
+             Feedback.msg_debug Pp.(str"abstracted type: " ++ prc evenv (nf_evar !evdref rhs'));
+           Evarsolve.check_evar_instance !evdref evk rhs'
+                                         (conv_fun evar_conv_x (default_flags_of full_transparent_state))
+         with IllTypedInstance _ -> raise (TypingFailed evd)
+       in
+      Evd.define evk rhs' evd
   in
-  abstract_free_holes evd subst, true
+  let evd = abstract_free_holes evd subst in
+  evd, true
   with TypingFailed evd -> evd, false
 
-let second_order_matching_with_args ts env evd pbty ev l t =
-(*
-  let evd,ev = evar_absorb_arguments env evd ev l in
-  let argoccs = Array.map_to_list (fun _ -> None) (snd ev) in
-  let evd, b = second_order_matching ts env evd ev argoccs t in
-  if b then Success evd
-  else UnifFailure (evd, ConversionFailed (env,mkApp(mkEvar ev,l),t))
-  if b then Success evd else
- *)
-  let pb = (pbty,env,mkApp(mkEvar ev,l),t) in
-  UnifFailure (evd, CannotSolveConstraint (pb,ProblemBeyondCapabilities))
+let default_evar_selection flags evd (ev,args) =
+  let evi = Evd.find_undefined evd ev in
+  let rec aux args abs =
+    match args, abs with
+    | _ :: args, a :: abs ->
+       let spec = if not flags.allow_K_at_toplevel then
+                    AtOccurrences (if a then Locus.AtLeastOneOccurrence else Locus.AllOccurrences)
+                  else Unspecified a in
+       spec :: aux args abs
+    | l, [] -> List.map (fun _ -> default_occurrence_selection) l
+    | [], _ :: _ -> assert false
+  in aux (Array.to_list args) evi.evar_abstract_arguments
+
+let second_order_matching_with_args flags env evd with_ho pbty ev l t =
+  if with_ho then
+    let evd,ev = evar_absorb_arguments env evd ev (Array.to_list l) in
+    let argoccs = default_evar_selection flags evd ev in
+    let test = default_occurrence_test ~frozen_evars:flags.frozen_evars flags.subterm_ts in
+    let evd, b =
+      try second_order_matching flags env evd ev (test,argoccs) t
+      with PretypeError (_, _, NoOccurrenceFound _) -> evd, false
+    in
+    if b then Success evd
+    else
+      UnifFailure (evd, ConversionFailed (env,mkApp(mkEvar ev,l),t))
+  else
+    let pb = (pbty,env,mkApp(mkEvar ev,l),t) in
+    UnifFailure (evd, CannotSolveConstraint (pb,ProblemBeyondCapabilities))
 
-let apply_conversion_problem_heuristic ts env evd pbty t1 t2 =
-  let t1 = apprec_nohdbeta ts env evd (whd_head_evar evd t1) in
-  let t2 = apprec_nohdbeta ts env evd (whd_head_evar evd t2) in
+let is_beyond_capabilities = function
+  | CannotSolveConstraint (pb,ProblemBeyondCapabilities) -> true
+  | _ -> false
+
+(* TODO frozen *)
```

Comment:
Right. What remained to do was to handle frozen evars correctly here, I just implemented it and removed the todo. Frozen evars cannot be instantiated so this was just a matter of restricting the rules so that no frozen evars is instantiated.